### PR TITLE
Document counter sanity checks and enforced delays

### LINKED ISSUE #106: Document sanity checks
When I did #44, I forgot to document the noise weight sanity checks in a doc/ directory.
It might be best to leave this until #9, #22, and #75 are completed, because the checks may change. (And maybe also #109 and #111.)

Here are the current checks from #44:
* the DC threshold is between 1 and privcount.util.MAX_DC_COUNT (currently 1 million)
  * there must be at least 1 DC to count anything
  * MAX_DC_COUNT is a sanity check against typos and similar errors 
* each noise weight is >= 0.0
  * negative noise weights do not make sense, as the noise values are already symmetrically distributed around 0.0
* the total noise weight is <= max_tally_counter_value() (currently 590295810358705651711 ~ `5.9 * 10^20` with privcount.util.counter_modulus() at `2**70`)
  * this is a sanity check against adding more noise than the counter maximum (which makes the counter meaningless, and is likely to be a software bug or misconfiguration)

### LINKED ISSUE #127: Document PrivCount enforced delays
Split off #22.

From the config file:
* delay_period: 4 # (default: collect_period) the number of seconds the TS should wait between rounds that change noise allocations
* continue: 2 # start another collection phase after finishing a previous collection phase. If this value is an integer, run that many rounds before stopping. (The TS always runs at least 1 round.)

From the code:
            # The delay period must be greater than or equal to the collect     
            # period

            # we want the end time after all clients have definitely stopped    
            # and returned their results, not the time the TS told the          
            # CollectionPhase to initiate the stop procedure                    
            # (otherwise, a lost message or down client could delay stopping    
            # for a pathological period of time, breaking our assumptions)      
            # This also means that the SKs will allow the round to start        
            # slightly before the TS allows it, which is a good thing.       

class CollectionDelay(object):
    '''                                                                         
    Ensures a configurable delay between rounds with different noise            
    allocations.                                                                
    Usage:
    (all nodes do these checks, but the nodes listed must enforce them)         
    TS: configures round                                                        
        uses get_next_round_start_time() for status updates                     
        checks round_start_permitted() before starting collection               
    TODO: DC: checks round_start_permitted() before sending blinding shares
    TODO: SK: checks round_start_permitted() before accepting blinding shares   
    (round runs)                                                                
    TODO: SK: set_stop_result() when round stops and blinding shares are sent   
    TODO: DC: set_stop_result() when round stops and counters are sent
    TS: set_stop_result() when round ends successfully                          
    (repeat for next round, if TS has continue set in its config)

    def sigma_change_needs_delay(previous_sigma, proposed_sigma,
                                 logging_label=None):
        '''                                                                     
        Check if there should be a delay between rounds using the previous      
        and proposed sigma values for the same counter.                         
        A counter can use two sigma values without a delay between them if:     
        - The values are equal, or                                              
        - The proposed value is greater than the previous value.                
        TODO: Allow small decreases within a configurable tolerance.

    def noise_change_needs_delay(previous_allocation, proposed_allocation):
        '''                                                                     
        Check if there should be a delay between rounds using the previous      
        and proposed noise allocations.                                         
        Two allocations can be used without a delay between them if:            
        - They have the same keys, and                                          
        - The sigma values for those keys do not need a delay.                  
        Returns True if the allocations need a delay, False if they do not.     
        '''

   def set_stop_result(self, noise_allocation, start_time, end_time, delay,
                        round_successful):
        '''                                                                     
        Called when a round ends.                                               
        If the new noise allocation is not equivalent to the stored noise,      
        update the stored noise. Update the stored last round end time.         
        No updates are performed for failed rounds.                             
        '''

                # The latest noise allocation could have been used immediately  
                # after the starting noise allocation.                          
                # Keep the starting noise allocation, so that a TS can't        
                # gradually decrease the noise each round

### PR DESCRIPTION
**Pull request by teor2345**
Closes #106.
Closes #127.

### CODE REVIEW

**Comment by teor2345** [comment]
Given the size of this document, I'd like someone else to read it before it gets merged.

File: doc/CounterChecks.markdown [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+* The upper bound is strictly greater than the lower bound
+* The bins in a counter must not overlap (but there may be gaps)
+
+### Periods
+
+The periods determine how often nodes communicate, and how long rounds last.
+
+#### Collect Period
+
+Data Collectors collect events for collect period seconds.
```
Comment (robgjansen) [inline doc/CounterChecks.markdown]: I would add that the events are collected from Tor and that PrivCount increments the counters corresponding to the events it receives.

File: doc/CounterChecks.markdown [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+tally_server:
+  collect_period: 86400
+```
+
+* the collect period must be a non-negative integer
+* a collect period shorter than 4 seconds does not work reliably
+
+#### Event Period
+
+The tally server checks for new events every event period.
```
Comment (robgjansen) [inline doc/CounterChecks.markdown]: I'm not sure what checking for new events means... After checking the code, this is how often the tally server runs its main event loop to do general maintenance tasks like check for config file changes, check on DC and SK status, check if a round ended, etc.

File: doc/CounterChecks.markdown [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  event_period: 600
+```
+
+* the event period must be a non-negative integer
+* there must be at least two event periods in each collect period
+* event periods shorter than 2 seconds do not work reliably
+
+#### Checkin Period
+
+Data Collectors and Share Keepers check in every checkin period.
```
Comment (robgjansen) [inline doc/CounterChecks.markdown]: I would clarify what "check in" means.

File: doc/CounterChecks.markdown [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  delay_period: 86400
+```
+
+* the delay period must be a non-negative integer
+* PrivCount warns if the delay period is very short
+
+### Node Thresholds
+
+The node thresholds determine when rounds are run.
+The Tally Server does not start the round until both node thresholds are met.
```
Comment (robgjansen) [inline doc/CounterChecks.markdown]: i.e., when the number of nodes of each type that have joined our deployment exceeds the configured thresholds.

**Comment by teor2345** [comment]
I am happy with how this looks now. Merging.

**Pull request merged**